### DISCUSSION TITLE
Better engine detection

### DISCUSSION THREAD

**Discussion by Neved4** [tectonic-typesetting/tectonic#958] [Ideas]
LaTeX3 has some convenient tools for package writers. In `l3kernel -> l3sys` we have `\c_sys_engine_str`, that holds the name of the engine (luatex, pdftex, xetex... etc); and `\sys_if_engine_xetex_p:`, a boolean that is true on XeTeX-based engines.  
In LaTeX2 world, people resort to the `iftex` package.

As `tectonic` evolves, it'll differentiate itself from XeTeX. Having a way to detect XeTeX-based is nice, and I'm not aware if we have a way of detecting `tectonic` unequivocally. That would come handy for situations where writing `tectonic`-specific conditional code is needed.

Part of this is outside the project direct control: both LaTeX3 and other packages would need to add options to detect `tectonic`, but we could keep this issue to keep tabs on those.

Internally the question is about how we could bridge that detection of `tectonic`: an internal macro only present when compiled with the engine? That would suffice for developers, regardless of other packages adding `tectonic` to their features list.

If suitable, I'm open to converting this into a Discussion.

**Comment by pkgw** [comment]
Yup, this is something to think about! I have a couple of quick comments. For the foreseeable future, I think that engine selectors like `\c_sys_engine_str` should report whatever they do for XeTeX: I don't want to force package authors to have to write the equivalent of `\ifxetex || \iftectonic` everywhere.

Also, Tectonic does provide a new primitive called `\TectonicCodaTokens` that could be used to implement a hypothetical `\iftectonic`.
